interfaces/cython/cantera/composite.py

```diff
@@ -528,32 +528,47 @@ def __init__(self, phase, shape=(0,), states=None, extra=None, meta=None):
         reserved = self.__dir__()
 
         self._extra = OrderedDict()
+
+        if isinstance(extra, str):
+            extra = [extra]
+
         if isinstance(extra, dict):
             for name, v in extra.items():
                 if name in reserved:
                     raise ValueError(
                         "Unable to create extra column '{}': name is already "
                         "used by SolutionArray objects.".format(name))
                 if not np.shape(v):
-                    self._extra[name] = [v]*self._shape[0]
-                elif len(v) == self._shape[0]:
-                    self._extra[name] = list(v)
+                    self._extra[name] = np.array([v]).reshape(self._shape)
+                elif len(v) == self._shape or np.array(v).shape == self._shape:
+                    self._extra[name] = np.array(v)
                 else:
                     raise ValueError("Unable to map extra SolutionArray"
                                      "input for named {!r}".format(name))
 
-        elif extra and self._shape == (0,):
+        elif extra is not None:
+            try:
+                iterator = iter(extra)
+            except TypeError :
+                raise ValueError(
+                    "Extra properties can be created by passing an iterable"
+                    " of names for the properties, if the SolutionArray is not initially"
+                    " empty. If you want to supply initial values for the properties, use"
+                    " a dictionary whose keys are the names of the properties and values "
+                    " are the initial values.") from None
+
             for name in extra:
```

Comment:
You might as well use the iterator you created here, probably best to call it something more specific, like `extra_iter` or something like that.